### PR TITLE
[Form] Use `form.post_set_data` in `ResizeFormListener`

### PR DESCRIPTION
**Pull request by HeahDude**
| Q             | A
| ------------- | ---
| Branch?       | 7.2
| Bug fix?      | no
| New feature?  | yes
| Deprecations? | yes
| Tickets       | ~
| License       | MIT
| Doc PR        | TODO

The `form.pre_set_data` event allows to change the initial data, modifying the form based on it should be done early during `form.post_set_data` instead.

Ref https://github.com/symfony/symfony-docs/pull/18587.

Also, the collect of form config should be done as late as possible to account for modification from other listeners.

### CODE REVIEW

File: src/Symfony/Component/Form/Extension/DataCollector/EventListener/DataCollectorListener.php [hunk 1]
```diff
@@ -34,8 +34,8 @@ public function __construct(FormDataCollectorInterface $dataCollector)
     public static function getSubscribedEvents(): array
     {
         return [
-            // High priority in order to be called as soon as possible
-            FormEvents::POST_SET_DATA => ['postSetData', 255],
+            // Low priority in order to be called as late as possible
```
Comment (xabbuh) [inline src/Symfony/Component/Form/Extension/DataCollector/EventListener/DataCollectorListener.php:35]: Do you know why the priority was really high before? Could it be that we will forbid some use case with these changes?

Comment (HeahDude) [reply src/Symfony/Component/Form/Extension/DataCollector/EventListener/DataCollectorListener.php:35]: Not really, this is there since the origin (#9082).
There are no core usages of post set data event (except for a recent deprecation concerning timezone :)).
But the changes in this PR tells me that if any listener out there modifies the form structure, we would collect a tree out of sync and may display irrelevant information in the profiler.

**Review by xabbuh — APPROVED** [review]

**Review by yceruto — APPROVED** [review]

File: src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php [hunk 2]
```diff
@@ -77,6 +80,11 @@ public function preSetData(FormEvent $event)
         }
     }
 
+    public function postSetData(PostSetDataEvent $event): void
+    {
+        $this->preSetData($event);
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php]: this should be done the other way around: the deprecated method calls the non-deprecated one

File: src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php [hunk 3]
```diff
@@ -45,14 +46,16 @@ public function __construct(string $type, array $options = [], bool $allowAdd =
...
             FormEvents::SUBMIT => ['onSubmit', 50],
         ];
     }
 
     /**
+     * @deprecated Since 6.4, use {@see postSetData()} instead.
+     *
      * @return void
      */
     public function preSetData(FormEvent $event)
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php]: missing runtime deprecation in the body of the method

File: src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php (hunk 2 above)
Comment (HeahDude) [reply src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php]: By doing the other way around we might break a child class by not calling `preSetData` anymore.
The last time I did such deprecation we handle it that way https://github.com/symfony/symfony/commit/ad08b951d798a5419f57bb092d636ff2c86dfece, https://github.com/symfony/symfony/commit/8a6cf9d93eda5d4c7685fd794a3733f787368055, should we do something similar instead?

Comment (nicolas-grekas) [reply src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php]: I guess so yes. The runtime notice is a must.

Comment (HeahDude) [reply src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php]: See https://github.com/symfony/symfony/pull/51041/commits/ba36dd37091dd6df8b7953185c4bfc0f1f1bf920.

File: src/Symfony/Component/Form/CHANGELOG.md [hunk 4]
```diff
@@ -8,6 +8,8 @@ CHANGELOG
    `model_timezone` option in `DateType`, `DateTimeType`, and `TimeType`
  * Deprecate `PostSetDataEvent::setData()`, use `PreSetDataEvent::setData()` instead
  * Deprecate `PostSubmitEvent::setData()`, use `PreSubmitDataEvent::setData()` or `SubmitDataEvent::setData()` instead
+ * Use `form.post_set_data` instead of `form.pre_set_data` in `ResizeFormListener`
+ * Change the priority of `DataCollectorListener` from 255 to -255
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/CHANGELOG.md]: Are we sure about this? From experience this can lead to unwanted behaviors, eg what if a listener stop propagation? (is that possible with Doctrine?)
What's the motivation for this change of priority? Is that just modern-us thinking ancient-us made a mistake? Given the comment, this looks on purpose to me.

File: src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php [hunk 5]
```diff
@@ -45,18 +50,51 @@ public function __construct(string $type, array $options = [], bool $allowAdd =
...
+            || __CLASS__ === (new \ReflectionClass($this))->getMethod('preSetData')->getDeclaringClass()->name
+        ) {
+            // not a child class, or child class does not overload PRE_SET_DATA
+        } else {
+            trigger_deprecation('symfony/form', '6.4', 'Calling "%s()" is deprecated, use "%s::postSetData()" instead.', __METHOD__, __CLASS__);
+            // parent::preSetData() has been called
+            $this->overridden = false;
+            $this->postSetData($event);
+            $this->usePreSetData = true;
+        }
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php]:
```suggestion
            // not a child class, or child class does not overload PRE_SET_DATA
            return;
        }
        trigger_deprecation('symfony/form', '6.4', 'Calling "%s()" is deprecated, use "%s::postSetData()" instead.', __METHOD__, __CLASS__);
        // parent::preSetData() has been called
        $this->overridden = false;
        $this->postSetData($event);
        $this->usePreSetData = true;
```

File: src/Symfony/Component/Form/CHANGELOG.md (hunk 4 above)
Comment (HeahDude) [reply src/Symfony/Component/Form/CHANGELOG.md]: It raises the same concern as @xabbuh in https://github.com/symfony/symfony/pull/51041#discussion_r1270524066.
Yes, I do think that was a mistake.
Given than we don't currently use post set data event in core, I cannot think of a legit use case where propagation is stopped.
To ensure that this listener is called, we could use a flag and `__destruct()` to warn or deprecate such behavior.

Comment (nicolas-grekas) [reply src/Symfony/Component/Form/CHANGELOG.md]: OK (no need for the destruct check)

File: src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php (hunk 5 above)
Comment (HeahDude) [reply src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php]: Done.

File: src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php [hunk 6]
```diff
@@ -45,18 +50,52 @@ public function __construct(string $type, array $options = [], bool $allowAdd =
     public static function getSubscribedEvents(): array
     {
         return [
-            FormEvents::PRE_SET_DATA => 'preSetData',
+            FormEvents::PRE_SET_DATA => 'preSetData', // deprecated
+            FormEvents::POST_SET_DATA => ['postSetData', 255], // as early as possible
```
Comment (stof) [inline src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php:52]: is it actually allowed to modify the structure of the form on postSetData ? To me, this is too late as those child fields won't be populated.

Comment (HeahDude) [reply src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php:52]: It is, the form is not cloned and can be modified by reference https://github.com/symfony/symfony/blob/6.4/src/Symfony/Component/Form/Form.php#L327.
The only event that forbids it is `POST_SUBMIT` thanks to this check https://github.com/symfony/symfony/blob/6.4/src/Symfony/Component/Form/Form.php#L260, in such case `$this->defaultDataSet` is true, but not enough, only data cannot further be changed.

Comment (HeahDude) [reply src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php:52]: >To me, this is too late as those child fields won't be populated.

They will be: https://github.com/symfony/symfony/blob/6.4/src/Symfony/Component/Form/Form.php#L776.

Comment (HeahDude) [reply src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php:52]: I've refactored the tests to ensure data is mapped consistently: https://github.com/symfony/symfony/pull/51041/commits/6588362b4f6a72b0b956b18aa46a8e52a39a5dce.

**Comment by OskarStark** [comment]
Needs a rebase, thanks

**Comment by nicolas-grekas** [comment]
Friendly up @HeahDude

**Comment by xabbuh** [comment]
I have rebased the PR

File: src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php [hunk 7]
```diff
@@ -44,15 +48,50 @@ public function __construct(
...
+        ) {
+            // not a child class, or child class does not overload PRE_SET_DATA
+            return;
+        }
+
+        trigger_deprecation('symfony/form', '7.2', 'Calling "%s()" is deprecated, use "%s::postSetData()" instead.', __METHOD__, __CLASS__);
+        // parent::preSetData() has been called
+        $this->overridden = false;
+        $this->postSetData($event);
+        $this->usePreSetData = true;
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php]:
```suggestion
        try {
            $this->postSetData($event);
        } finally {
            $this->usePreSetData = true;
        }
```

File: src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php [hunk 8]
```diff
@@ -44,15 +48,50 @@ public function __construct(
...
+    }
+
+    // Remove FormEvent type hint in 8.0
+    final public function postSetData(FormEvent|PostSetDataEvent $event): void
```
Comment (nicolas-grekas) [inline src/Symfony/Component/Form/Extension/Core/EventListener/ResizeFormListener.php]: `@final since Symfony 7.3` insead
```suggestion
    /**
     * @param PostSetDataEvent $event
     */
    public function postSetData(FormEvent|PostSetDataEvent $event): void
```

**Review by nicolas-grekas — APPROVED** [review]
LGTM after minor comments.

**Comment by xabbuh** [comment]
Thank you Jules.

**Comment by HeahDude** [comment]
Thank you Christian :).
